lambdas/pkgpush/index.py

```diff
@@ -172,6 +179,50 @@
 logger = get_quilt_logger()
 
 
+def calculate_pkg_hashes(s3_client, pkg):
+    entries = []
+    for lk, entry in pkg.walk():
+        if entry.hash is not None:
+            continue
+        if entry.size > S3_HASH_LAMBDA_MAX_FILE_SIZE:
+            raise FileTooLargeForHashing(lk)
+
+        entries.append(entry)
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=S3_HASH_LAMBDA_CONCURRENCY) as pool:
+        fs = [
+            pool.submit(calculate_pkg_entry_hash, s3_client, entry)
+            for entry in entries
+        ]
+        for f in concurrent.futures.as_completed(fs):
+            f.result()
+
+
+def calculate_pkg_entry_hash(s3_client, pkg_entry):
+    pk = pkg_entry.physical_key
+    params = {
+        'Bucket': pk.bucket,
+        'Key': pk.path,
+    }
+    if pk.version_id is not None:
+        params['VersionId'] = pk.version_id
+    url = s3_client.generate_presigned_url(
+        ClientMethod='get_object',
+        ExpiresIn=S3_HASH_LAMBDA_SIGNED_URL_EXPIRES_IN,
+        Params=params,
+    )
+    pkg_entry.hash = {
+        'type': 'SHA256',
+        'value': invoke_hash_lambda(url),
```

Comment:
Are we sure that one file per lambda is the fastest? What about 2 or 3 or N?

Reply:
N is what we do now without dedicated hashing lambda.

There is no "optimal" predefined number for this as it depends on number of files, their sizes and lambda's remaining time. I don't want to add complexity with implementing such scheduler and making hash lambda work with many files in parallel because current implementation seem to work OK.